[AutoDiff] Emit a diagnostic for non-differentiable active values

asavonic opened the pull request:
For some values we cannot compute types for differentiation (for example, tangent vector type), so it is better to diagnose them earlier. Otherwise we hit assertions when generating code for such invalid values.

The LIT test is a reduced reproducer from the issue #66996. Before the patch the compiler crashed while trying to get a tangent vector type for the following value (partial_apply):

    %54 = function_ref @$s4null1o2ffSdAA1FV_tFSdyKXEfu0_ :
      $@convention(thin) @substituted <τ_0_0> (@inout_aliasable Double)
      -> (@out τ_0_0, @error any Error) for <Double>

    %55 = partial_apply [callee_guaranteed] %54(%2) :
      $@convention(thin) @substituted <τ_0_0> (@inout_aliasable Double)
      -> (@out τ_0_0, @error any Error) for <Double>

Now we emit a diagnostic instead.

The patch resolves issues #66996 and #63331.

asavonic committed:
[AutoDiff] Emit a diagnostic for non-differentiable active values

For some values we cannot compute types for differentiation (for example,
tangent vector type), so it is better to diagnose them earlier. Otherwise we hit
assertions when generating code for such invalid values.

The LIT test is a reduced reproducer from the issue #66996. Before the patch the
compiler crashed while trying to get a tangent vector type for the following
value (partial_apply):

    %54 = function_ref @$s4null1o2ffSdAA1FV_tFSdyKXEfu0_ :
      $@convention(thin) @substituted <τ_0_0> (@inout_aliasable Double)
      -> (@out τ_0_0, @error any Error) for <Double>

    %55 = partial_apply [callee_guaranteed] %54(%2) :
      $@convention(thin) @substituted <τ_0_0> (@inout_aliasable Double)
      -> (@out τ_0_0, @error any Error) for <Double>

Now we emit a diagnostic instead.

The patch resolves issues #66996 and #63331.

asl commented:
Tagging @BradLarson

asl commented:
@swift-ci please test

rxwei commented on lib/SILOptimizer/Differentiation/PullbackCloner.cpp:
```diff
@@ -1804,6 +1804,11 @@ bool PullbackCloner::Implementation::run() {
   DominanceOrder domOrder(original.getEntryBlock(), domInfo);
   // Keep track of visited values.
   SmallPtrSet<SILValue, 8> visited;
+  // Also check for differentiable for all active values
```
Not sure I understand this comment. The two lines of code below are just obtaining the protocol and the module and don't seem related to the comment.

asavonic replied:
Right, this comment is more suitable next to the actual check below.

asavonic committed:
Move comment next to the check

asl commented:
@swift-ci please test

asavonic commented:
There is quite a lot of LIT test with tuples being incorrectly (?) diagnosed as non-differentiable. I will check them and update the patch.

asl commented:
@asavonic Because some components are non-differentiable?

asavonic committed:
Check getTangentSpace instead of checking for conformance to the protocol

Types like `tuple` do not conform to the differentiable protocol, but they are
handled in the compiler as special cases.

getTangentSpace should work for all these types, and discard truly
non-differentiable types.

asavonic committed:
Use 'operator bool' instead of 'has_value'

asl commented:
@swift-ci please test

asl commented:
@shahmishal any idea why CI is not triggered? (see https://github.com/apple/swift/pull/67697#issuecomment-1679437979)

rxwei reviewed (approved):
Sorry for the delayed review
